system/mp/mpi_wrapper.cpp

```diff
@@ -0,0 +1,513 @@
+    CriticalBlock block(commDataLock);
+    int size = asyncCommData.size();
+    for(int i=(size-1); i>=0; i--)
+    {
+        if (asyncCommData[i]==commData)
+        {
+            _popCommData(i);
+            break;
+        }
+    }
+    return commData;
+}
+
+//----------------------------------------------------------------------------------------------------------------------------//
+
+#define TAG_UB 100000 //INT_MAX //MPI_TAG_UB
+int getRank(rank_t sourceRank)
+{
+    _TF("getRank", sourceRank);
+    if (sourceRank == RANK_ALL)
+        return MPI::ANY_SOURCE;
+    else
+        return sourceRank;
+}
+
+int getTag(mptag_t mptag)
+{
+    _TF("getTag", mptag);
+    int returnVal;
+    unsigned tag = mptag;
+    if (mptag == TAG_ALL)
+        returnVal = MPI::ANY_TAG;
+    else
+    {
+        if (tag > TAG_UB)
+        {
+            tag= TAG_UB-((unsigned)-tag);                         //MPI doesn't allow custom tags with negative values or beyond MPI_TAG_UB. Thus we shift it to the MPI_TAG_UB range
+        }
+        returnVal = static_cast<int>(tag);
+
+    }
+    return returnVal;
+}
+
+mptag_t getTag(int tag)
+{
+    _TF("getTag", tag);
+    unsigned returnVal;
+    if (tag == MPI::ANY_TAG)
+        returnVal = TAG_ALL;
+    else
+    {
+        if (tag > (TAG_UB/2))
+        {
+            returnVal = (unsigned) - (TAG_UB-tag);                         //MPI doesn't allow custom tags with negative values
+        } else
+            returnVal = tag;
+
+    }
+    return (mptag_t)returnVal;
+}
+
+MPI::Status waitToComplete( bool& completed, bool& error, bool& canceled, bool& timedout, unsigned timeout, CommData *commData)
+{
+    _TF("waitToComplete", completed, error, canceled, timeout);
+    CTimeMon tm(timeout);
+    MPI::Status stat;
+    unsigned remaining;
+    bool noCancellation;
+    while ((noCancellation = commData->lockFromCancellation()) && !(completed || error || (timedout = tm.timedout(&remaining))))
+    {
+        completed = commData->request.Test(stat);
+        commData->releaseCancellationLock();
+        usleep(WAIT_DELAY);
+    }
+    if (completed)
+        canceled = stat.Is_cancelled();
+    else
+        canceled = !noCancellation;
+
+    return stat;
+}
+
+MPI::Status hasIncomingData(int sourceRank, int mptag, MPI::Comm& comm,
+        bool& incomingMessage, bool& error, bool& canceled, unsigned timeout, CommData *commData)
+{
+    _TF("hasIncomingData", sourceRank, mptag, incomingMessage, error, timeout);
+
+    CTimeMon tm(timeout);
+    MPI::Status stat;
+
+    int flag;
+    unsigned remaining;
+    bool noCancellation;
+
+    while ((noCancellation = commData->lockFromCancellation()) && !(incomingMessage || error || (timeout !=0 && tm.timedout(&remaining))))
+    {
+        incomingMessage = comm.Iprobe(sourceRank, mptag, stat);
+        commData->releaseCancellationLock();
+        if (timeout == 0) break;
+        usleep(PROBING_DELAY);
```

Comment:
this looks like a tight loop (albeit with a delay).
Will need to revisit this paradigm to look at blocking and be signalling instead.

Reply:
same reply as above.